File: backend/services/communityService.js
const db = require('../config/db');

exports.createChannel = async (creatorId, channelData) => {
  console.log('[CommunityService] Criando canal:', channelData.name);

  try {
    const [result] = await db.query(
      'INSERT INTO channels (creator_id, name, description, type, is_private, tier_required) VALUES (?, ?, ?, ?, ?, ?)',
      [creatorId, channelData.name, channelData.description, channelData.type, channelData.isPrivate || false, channelData.tierRequired]
    );
    return { id: result.insertId, ...channelData, creatorId };
  } catch (err) {
    console.error('[CommunityService] Erro ao criar canal:', err);
    throw err;
  }
};

exports.getChannels = async (userId) => {
  console.log('[CommunityService] Buscando canais para usuário:', userId);

  try {
    // Buscar canais públicos ou onde o usuário é membro
    const [rows] = await db.query(
      'SELECT * FROM channels WHERE is_private = FALSE OR creator_id = ? OR JSON_CONTAINS(members, ?)',
      [userId, JSON.stringify(userId)]
    );
    return rows.map(row => ({
      id: row.id,
      creatorId: row.creator_id,
      name: row.name,
      description: row.description,
      type: row.type,
      isPrivate: row.is_private,
      tierRequired: row.tier_required,
      members: JSON.parse(row.members || '[]'),
      createdAt: row.created_at
    }));
  } catch (err) {
    console.error('[CommunityService] Erro ao buscar canais:', err);
    throw err;
  }
};

exports.joinChannel = async (userId, channelId) => {
  console.log('[CommunityService] Usuário juntando-se ao canal:', userId, channelId);

  try {
    // Verificar se canal existe e é acessível
    const [channels] = await db.query('SELECT * FROM channels WHERE id = ?', [channelId]);
    if (channels.length === 0) {
      throw new Error('Canal não encontrado');
    }
    const channel = channels[0];

    // Verificar acesso ao tier se privado
    if (channel.is_private && channel.tier_required) {
      const [supporters] = await db.query('SELECT id FROM tier_supporters WHERE user_id = ? AND tier_id = ?', [userId, channel.tier_required]);
      if (supporters.length === 0) {
        throw new Error('Acesso negado: tier necessário não suportado');
      }
    }

    // Adicionar usuário aos membros
    const members = JSON.parse(channel.members || '[]');
    if (!members.includes(userId)) {
      members.push(userId);
      await db.query('UPDATE channels SET members = ? WHERE id = ?', [JSON.stringify(members), channelId]);
    }

    return true;
  } catch (err) {
    console.error('[CommunityService] Erro ao juntar-se ao canal:', err);
    throw err;
  }
};

exports.sendMessage = async (senderId, channelId, messageData) => {
  console.log('[CommunityService] Enviando mensagem:', senderId, channelId);

  try {
    // Verificar acesso ao canal
    const [channels] = await db.query('SELECT * FROM channels WHERE id = ?', [channelId]);
    if (channels.length === 0) {
      throw new Error('Canal não encontrado');
    }
    const channel = channels[0];

    // Verificar se usuário é membro ou criador
    const members = JSON.parse(channel.members || '[]');
    if (channel.creator_id !== senderId && !members.includes(senderId)) {
      throw new Error('Acesso negado ao canal');
    }

    // Verificar tier se mensagem privada
    if (messageData.isPrivate && messageData.tierRequired) {
      const [supporters] = await db.query('SELECT id FROM tier_supporters WHERE user_id = ? AND tier_id = ?', [senderId, messageData.tierRequired]);
      if (supporters.length === 0) {
        throw new Error('Acesso negado: tier necessário não suportado');
      }
    }

    const [result] = await db.query(
      'INSERT INTO messages (sender_id, channel_id, text, is_private, tier_required) VALUES (?, ?, ?, ?, ?)',
      [senderId, channelId, messageData.text, messageData.isPrivate || false, messageData.tierRequired]
    );

    return {
      id: result.insertId,
      senderId,
      channelId,
      text: messageData.text,
      timestamp: new Date(),
      isPrivate: messageData.isPrivate || false,
      tierRequired: messageData.tierRequired
    };
  } catch (err) {
    console.error('[CommunityService] Erro ao enviar mensagem:', err);
    throw err;
  }
};

exports.getMessages = async (channelId, limit = 50) => {
  console.log('[CommunityService] Buscando mensagens do canal:', channelId);

  try {
    const [rows] = await db.query(
      'SELECT m.*, u.name as sender_name FROM messages m JOIN users u ON m.sender_id = u.id WHERE m.channel_id = ? ORDER BY m.timestamp DESC LIMIT ?',
      [channelId, limit]
    );
    return rows.map(row => ({
      id: row.id,
      senderId: row.sender_id,
      senderName: row.sender_name,
      channelId: row.channel_id,
      text: row.text,
      timestamp: row.timestamp,
      isPrivate: row.is_private,
      tierRequired: row.tier_required
    })).reverse(); // Ordem cronológica
  } catch (err) {
    console.error('[CommunityService] Erro ao buscar mensagens:', err);
    throw err;
  }
};

exports.createMuralPost = async (creatorId, channelId, postData) => {
  console.log('[CommunityService] Criando post de mural:', creatorId, channelId);

  try {
    // Verificar acesso ao canal
    const [channels] = await db.query('SELECT * FROM channels WHERE id = ?', [channelId]);
    if (channels.length === 0) {
      throw new Error('Canal não encontrado');
    }
    const channel = channels[0];

    // Verificar se usuário é membro ou criador
    const members = JSON.parse(channel.members || '[]');
    if (channel.creator_id !== creatorId && !members.includes(creatorId)) {
      throw new Error('Acesso negado ao canal');
    }

    // Verificar tier se post privado
    if (postData.isPrivate && postData.tierRequired) {
      const [supporters] = await db.query('SELECT id FROM tier_supporters WHERE user_id = ? AND tier_id = ?', [creatorId, postData.tierRequired]);
      if (supporters.length === 0) {
        throw new Error('Acesso negado: tier necessário não suportado');
      }
    }

    const [result] = await db.query(
      'INSERT INTO mural_posts (creator_id, channel_id, title, description, images, parent_id, is_private, tier_required) VALUES (?, ?, ?, ?, ?, ?, ?, ?)',
      [creatorId, channelId, postData.title, postData.description, JSON.stringify(postData.images || []), postData.parentId, postData.isPrivate || false, postData.tierRequired]
    );

    return {
      id: result.insertId,
      creatorId,
      channelId,
      title: postData.title,
      description: postData.description,
      images: postData.images || [],
      parentId: postData.parentId,
      createdAt: new Date(),
      likes: 0,
      replies: [],
      isPrivate: postData.isPrivate || false,
      tierRequired: postData.tierRequired
    };
  } catch (err) {
    console.error('[CommunityService] Erro ao criar post de mural:', err);
    throw err;
  }
};

exports.getMuralPosts = async (channelId) => {
  console.log('[CommunityService] Buscando posts de mural do canal:', channelId);

  try {
    const [rows] = await db.query(
      'SELECT mp.*, u.name as creator_name FROM mural_posts mp JOIN users u ON mp.creator_id = u.id WHERE mp.channel_id = ? ORDER BY mp.created_at DESC',
      [channelId]
    );
    return rows.map(row => ({
      id: row.id,
      creatorId: row.creator_id,
      creatorName: row.creator_name,
      channelId: row.channel_id,
      title: row.title,
      description: row.description,
      images: JSON.parse(row.images || '[]'),
      parentId: row.parent_id,
      createdAt: row.created_at,
      likes: row.likes,
      replies: JSON.parse(row.replies || '[]'),
      isPrivate: row.is_private,
      tierRequired: row.tier_required
    }));
  } catch (err) {
    console.error('[CommunityService] Erro ao buscar posts de mural:', err);
    throw err;
  }
};

File: lib/services/community_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/community_models.dart';
import '../mock_data.dart';

/// Serviço para gerenciar funcionalidades de comunidade via API
class CommunityService {
  final String baseUrl;

  CommunityService({required this.baseUrl});

  /// Cria um novo canal
  /// @param creatorId ID do criador
  /// @param channelData Dados do canal
  /// @returns Canal criado
  Future<Channel> createChannel(String creatorId, Map<String, dynamic> channelData) async {
    final response = await http.post(
      Uri.parse('$baseUrl/community/users/$creatorId/channels'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(channelData),
    );

    if (response.statusCode == 201) {
      final data = jsonDecode(response.body);
      return Channel.fromJson(data);
    } else {
      throw Exception('Erro ao criar canal');
    }
  }

  /// Busca canais acessíveis ao usuário
  /// @param userId ID do usuário
  /// @returns Lista de canais
  Future<List<Channel>> getChannels(String userId) async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/community/users/$userId/channels'));

      if (response.statusCode == 200) {
        final List data = jsonDecode(response.body);
        return data.map((item) => Channel.fromJson(item)).toList();
      } else {
        throw Exception('Erro ao buscar canais');
      }
    } catch (e) {
      // Fallback para dados mock quando API não está disponível
      return mockChannels;
    }
  }

  /// Junta-se a um canal
  /// @param userId ID do usuário
  /// @param channelId ID do canal
  /// @returns Status da operação
  Future<void> joinChannel(String userId, String channelId) async {
    final response = await http.post(Uri.parse('$baseUrl/community/channels/$channelId/join/$userId'));

    if (response.statusCode != 200) {
      throw Exception('Erro ao juntar-se ao canal');
    }
  }

  /// Envia uma mensagem em um canal
  /// @param senderId ID do remetente
  /// @param channelId ID do canal
  /// @param messageData Dados da mensagem
  /// @returns Mensagem enviada
  Future<Message> sendMessage(String senderId, String channelId, Map<String, dynamic> messageData) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/community/channels/$channelId/messages/$senderId'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode(messageData),
      );

      if (response.statusCode == 201) {
        final data = jsonDecode(response.body);
        return Message.fromJson(data);
      } else {
        throw Exception('Erro ao enviar mensagem');
      }
    } catch (e) {
      // Fallback para modo mock: cria mensagem local e adiciona à lista mock
      final newMessage = Message(
        id: 'msg_${DateTime.now().millisecondsSinceEpoch}',
        senderId: senderId,
        channelId: channelId,
        text: messageData['text'] ?? '',
        timestamp: DateTime.now(),
        isPrivate: messageData['isPrivate'] ?? false,
        tierRequired: messageData['tierRequired'],
      );
      mockMessages.add(newMessage);
      return newMessage;
    }
  }

  /// Busca mensagens de um canal
  /// @param channelId ID do canal
  /// @param limit Limite de mensagens
  /// @returns Lista de mensagens
  Future<List<Message>> getMessages(String channelId, {int limit = 50}) async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/community/channels/$channelId/messages?limit=$limit'));

      if (response.statusCode == 200) {
        final List data = jsonDecode(response.body);
        return data.map((item) => Message.fromJson(item)).toList();
      } else {
        throw Exception('Erro ao buscar mensagens');
      }
    } catch (e) {
      // Fallback para dados mock quando API não está disponível
      return mockMessages.where((m) => m.channelId == channelId).toList();
    }
  }

  /// Cria um post de mural em um canal
  /// @param creatorId ID do criador
  /// @param channelId ID do canal
  /// @param postData Dados do post
  /// @returns Post criado
  Future<MuralPost> createMuralPost(String creatorId, String channelId, Map<String, dynamic> postData) async {
    final response = await http.post(
      Uri.parse('$baseUrl/community/channels/$channelId/posts/$creatorId'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(postData),
    );

    if (response.statusCode == 201) {
      final data = jsonDecode(response.body);
      return MuralPost.fromJson(data);
    } else {
      throw Exception('Erro ao criar post de mural');
    }
  }

  /// Busca posts de mural de um canal
  /// @param channelId ID do canal
  /// @returns Lista de posts
  Future<List<MuralPost>> getMuralPosts(String channelId) async {
    try {
      final response = await http.get(Uri.parse('$baseUrl/community/channels/$channelId/posts'));

      if (response.statusCode == 200) {
        final List data = jsonDecode(response.body);
        return data.map((item) => MuralPost.fromJson(item)).toList();
      } else {
        throw Exception('Erro ao buscar posts de mural');
      }
    } catch (e) {
      // Fallback para dados mock quando API não está disponível
      return mockMuralPosts.where((p) => p.channelId == channelId).toList();
    }
  }
}
